### ISSUE TITLE
Compile time constant expression evaluation

### ISSUE THREAD

**Issue by NunoFilipeSantos** [ethereum/solidity#13724]
### What

Evaluate constant expressions at compile time whenever possible.

### Resources

1. Older issues: https://github.com/ethereum/solidity/issues/3157

### Notes
Before starting, we need to wrap our heads around its interactions with #13365

### Related Issues:
- [ ] https://github.com/ethereum/solidity/issues/11436 - this will be a good progress benchmark for this at a later stage of implementation; also https://github.com/ethereum/solidity/issues/11972

**Comment by NunoFilipeSantos** [comment]
Closing in favor of #14284 (part of the development).

**Comment by 3esmit** [comment]
Is this feature still planned? Would it allow the use of library pure functions?

e.g.:
```solidity
    uint256 public constant A = 1000000;
    uint256 public constant B = 2000000;
    uint256 public constant C = 3000000;
    uint256 public constant D = Math.mulDiv(A, B, C);
 ```
